### ISSUE TITLE
Bug Report -- SystemError: <built-in function connect> returned NULL without setting an error

### ISSUE THREAD

**Issue by GrumpyB0T** [ibmdb/python-ibmdb#783]
I'm following the example noted here: https://www.ibm.com/docs/en/db2/11.5?topic=db-connecting-database-server

Creates the following error:

Traceback (most recent call last):
  File "XXXXXXXXXXXXXX.py", line 9, in <module>
    ibm_db_conn = ibm_db.connect(conn_str, '', '')
SystemError: <built-in function connect> returned NULL without setting an error

I believe there is a compatibility issue with the DB2 client libraries and ibm_db.  As long as the client instance version is below fixpack 2 of V11.1, the example code will work.  Additionally, no fixpack level of V11.5 seems to work.  The problem seems consistent between AIX and RHEL.

Example failing environment(s):
- Operating System Name: AIX 7300-00-01-2148
- Operating System Name: RHEL 8.6
- 
- db2level output: Informational tokens are "DB2 v11.5.4.0", "s2006161200", "DYN2006161200AIX", and Fix Pack "0".
- 
- Target Db2 Server Version: DB2/AIX64 10.5.10
-
- Python Version: Python 3.9.6 or 3.9.7
- 
- ibm_db version: ibm-db-3.1.3
- 
- Value of below environment variables if set:
- IBM_DB_HOME: /home/db2user/sqllib
- PATH: /opt/freeware/bin:/usr/bin:/etc:/usr/sbin:/usr/ucb:/usr/bin/X11:/sbin:/usr/java8_64/jre/bin:/usr/java8_64/bin
- LIB/LD_LIBRARY_PATH/DYLD_LIBRARY_PATH: LD_LIBRARY_PATH=$IBM_DB_HOME/lib64

**Comment by Earammak** [comment]
Can you please share the trace file.
Find the below steps:
cd clidriver install path, found with pip show ibm_db in Location: field of its output.
cd clidriver\bin
db2trc on -f trc.dmp
--> now run your application
db2trc off
db2trc flw trc.dmp trc.flw
db2trc fmt trc.dmp trc.fmt
db2trc fmt -c trc.dmp trc.fmtc
db2trc fmt -cli trc.dmp trc.cli
and share these files trc.flw, trc.fmt, trc.fmtc and trc.cli files

**Comment by GrumpyB0T** [comment]
Ran the trace, but the db2trc command was not found in Location path.  Had to load the profile for the client instance path instead.

[trc_flw.log](https://github.com/ibmdb/python-ibmdb/files/9418272/trc_flw.log)
[trc_fmt.log](https://github.com/ibmdb/python-ibmdb/files/9418273/trc_fmt.log)
[trc_fmtc.log](https://github.com/ibmdb/python-ibmdb/files/9418274/trc_fmtc.log)
[trc_cli.log](https://github.com/ibmdb/python-ibmdb/files/9418275/trc_cli.log)

**Comment by Earammak** [comment]
@GrumpyB0T ,  can you please share the output of "pip3 show ibm_db".

**Comment by GrumpyB0T** [comment]
Name: ibm-db
Version: 3.1.3
Summary: Python DBI driver for DB2 (LUW, zOS, i5) and IDS
Home-page: http://pypi.python.org/pypi/ibm_db/
Author: IBM Application Development Team
Author-email: praveen.narayanappa1@ibm.com
License: Apache License 2.0
Location: /opt/freeware/lib64/python3.9/site-packages
Requires: 
Required-by:

**Comment by imavo** [comment]
From the file "trc_cli.log", it appears that when python runs , python is picking up an antique version 10.1.0.0 cli driver c.f ": CLI Driver Version:        10.01.0000 ]. "

 Never run a fixpack 0 of any IBM software, there are always far too many bugs, always wait for the second or subsequent fixpacks if you value your time...otherwise you spend too much time resolving stilly bugs.


This can happen if the userid/account has dotted in an old client-instance db2profile (or an old server-instance db2profile if running locally on a Db2-server that has multiple Db2-instances).  Carefully check the shell startup files accessible to that userid to discover what is causing python to pick up the old version (e.g. position of directories on the PATH , or LD_LIBRARY_PATH etc).

If you no longer need a v10.1 client , then either uninstall that client-version, or upgrade it to match the version of your Db2-server (which the tracefile states is [ Product: QDB2/AIX64 DB2 v11.5.4.0 ] ), __OR__  at the very least modify the shell startup file(s) so that only the db2profile corresponding to your local Db2-server instance-version gets dotted in instead of the old one. Then retry python+ibm_db, after verifying that the tracefile shows the correct versions.

**Comment by GrumpyB0T** [comment]
The problem with this suggestion is the V10.1 product has never been installed on my test machine.  Additionally, V11.5.7.0 was the first version on AIX I tried before trying a series of back levels.

11.1.0.0 -- works
11.1.1.1 -- works
11.1.1.1a -- works
11.1.2.2 -- broken
11.5.4.0 -- broken
11.5.7.0 -- broken

Between each test the ibm_db module was uninstalled, the db2user client instance was dropped, and the pip cache directory was removed.

I had previous success years back on RHEL 8 with DB2 V11.1.0.0 and ibm_db 3.0.3, so I keep trying different DB2 versions on AIX until I found what seemed to be the breaking point.

**Comment by imavo** [comment]
There must be some other defect then because the cli trace mentions 10.01.0000  (in error, a false version number,  as you write no such version was previously installed on this hostname), but that is not a python or ibm_db problem.


You report the symptom also on RHEL, so I will check my RHEL environments, at least one has a local 11.5 version Db2 server so I can see if the current python ibm_db gives trouble.  I do not use AIX currently.

**Comment by Earammak** [comment]
@GrumpyB0T You can find the db2trc at /opt/freeware/lib64/python3.9/site-packages/clidriver/bin path in your system.
 you can share the trace files.

**Comment by imavo** [comment]
It seems this is not a general problem. Need to find out what is different on your RHEL 8.6 environment, and AIX environment.

For RHEL, I suggess carefully checking the output of `ldd` on the ibm_db shared object file.

I tested on RHEL 8.6 x64 with two local Db2-instances (one v11.5.7.0, other v10.5 ).

I used a separate python venv for each test.

All combinations of connecting with/without clidriver, to databases in both local Db2-instances, all worked successfully.

There was no occurrence of your symptom.

In all tests, I verified before installing ibm_db that both pip+setutools were at latest versions, and that the values of IBM_DB_HOME and LD_LIBRARY_PATH were appropriate for the test.

For tests without clidriver, i used the Db2-client that is part of the local Db2-server instances, setting IBM_DB_HOME/LD_LIBRARY_PATH appropriately.

For tests with clidriver, IBM_DB_HOME was not set when installing ibm_db, and LD_LIBRARY_PATH was exported with appropriate value before running the python connection test, and so I got 'todays' build of the clidriver downloaded each time (used --no-cache-dir).

I tested these combinations:

ibm_db 3.1.3 with clidriver - connecting to local v11.5.7.0 database
ibm_db 3.1.1 with clidriver - connecting to local v10.5 database
ibm_db 3.1.3 using v11.5.7.0 local client connecting to local v11.5.7.0 database
ibm_db 3.1.3 using v10.5 local client connecting to local v10.5 datatbase
ibm_db 3.1.3 using v10.5 local client connecting to local v11.5.7.0 database
ibm_db 3.1.3 using v11.5.7.0 local client connecting to local v10.5 database.

**Comment by Earammak** [comment]
@GrumpyB0T, Are you still facing the issue?  Can you please share the update on this.
Thanks,

**Comment by GrumpyB0T** [comment]
Tentatively solved for RHEL 8.6 on ppc64le.  However, setting IBM_DB_HOME and LD_LIBRARY_PATH was not enough, and I had to source everything in /home/db2user/sqllib/db2profile.  Adding all the equivalent environment changes to my own profile also worked.  As I found with AIX, only IBM_DB_HOME and LD_LIBRARY_PATH was required when staying under FP2 of v11.1.

Successful tested:
ibm_db 3.1.3 using v11.5.7.0 (Data Server Client V11.5 for Linux on Power) to remote v10.5 database
ibm_db 3.1.3 using v11.1.2.2 (Data Server Client V11.1 for Linux on Power) to remote v10.5 database


The same trick doesn't work for AIX, but I don't think this is directly the fault of the module or the DB2 client.  Messing with LIB paths doesn't seem to play well the freeware packages in general:

exec(): 0509-036 Cannot load program /opt/freeware/bin/python3.9 because of the 
following errors:
        0509-130 Symbol resolution failed for /opt/freeware/lib64/libpython3.9.so because:
        0509-136   Symbol _GLOBAL__AIXI_libintl_so (number 286) is not exported from
                   dependent module /usr/opt/rpm/lib/libintl.a[libintl.so.8].
        0509-136   Symbol _GLOBAL__AIXD_libintl_so (number 287) is not exported from
                   dependent module /usr/opt/rpm/lib/libintl.a[libintl.so.8].
        0509-192 Examine .loader section symbols with the
                 'dump -Tv' command.
